### DISCUSSION TITLE
How to detect when a an edge is dropped into the canvas? How to know which node/port has been dropped?

### DISCUSSION THREAD

**Discussion by Vadorequest** [reaviz/reaflow#52] [Q&A] [answered]
I want to create a node when dropping an edge into the canvas, but there doesn't seem to be any event notifying about this. It doesn't trigger `onNodeLinkCheck` or `onNodeLink `.

Example: https://youtu.be/S9Q91_sUT5A

Also, how can I know which node/port has been dropped? I assume it's stored somewhere, but I couldn't find how to get this information.

How can I achieve this?

**Comment by Vadorequest** [comment, marked as answer]
> Also, how can I know which node/port has been dropped? I assume it's stored somewhere, but I couldn't find how to get this information.

It's doable using `onDragStart` and `onDragEnd` as `Port` props.

See https://github.com/reaviz/reaflow/blob/147d64c4e8/src/symbols/Port/Port.tsx#L64-L65

I suppose `onDragEnd` can also be used to detect an edge is being dropped into the canvas? Gonna try that.

**Reply by Vadorequest** [reply]
Indeed, `onDragEnd` yields a `dragEvent` containing the native `event.target` which contains the DOM element being targeted.

![image](https://user-images.githubusercontent.com/3807458/107351959-8eb40580-6acb-11eb-892a-12a36d978d59.png)

```tsx

  const onDragEnd = (dragEvent: DragEvent, initial: Position, port: BasePortData, extra: any) => {
    const { xy, distance, event } = dragEvent;
    // @ts-ignore
    const { target } = event;
    console.log('onDragEnd port: ', node, dragEvent, initial, port, extra);
    console.log('at position', xy);
    console.log('draggedEdgeFromPort', draggedEdgeFromPort);
    console.log('target', target);

    if (onDragEndInternal) {
      onDragEndInternal(dragEvent, initial, port, extra);
    }
  };
```
 
The TS type for `event` seems wrong though, as it yields an error when trying to access `target` with `TS2339: Property 'target' does not exist on type 'UIEvent | React.UIEvent  | undefined'.` but works totally fine when the warning is being ignored.

![image](https://user-images.githubusercontent.com/3807458/107352050-b1461e80-6acb-11eb-81c4-705511aec1e7.png)

**Comment by amcdnl** [comment]
I use the events on the node and port for enter/leave and set them to a local state var.

**Reply by Vadorequest** [reply]
I see!

I think it's easier to use onDragEnd, as it contains everything that's needed (source node/port, event target), there is no need for state. (unless you want to know what's happening from another component during the drag)

Although, I might need enter/leave to easily resolve the node that's being targeted. But for detecting stuff that's dropped onto the canvas the above works flawlessly.
